### PR TITLE
deepcopy

### PR DESCRIPTION
**Pull request by hydroflame**
added deep copy

### CODE REVIEW

**Comment by hydroflame** [comment]
aahh, there might be a problem with string copy, if the original memory was RW, the copy isn't

**Comment by hydroflame** [comment]
I don't need 3 review, just one would be fine

File: deepcopy.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+		c.pointers[unsafe.Pointer(value.Pointer())] = s.Interface()
+
+		for i := 0; i < value.Len(); i++ {
+			s.Index(i).Set(reflect.ValueOf(c.deepCopy(value.Index(i).Interface())))
+		}
+
+		return s.Interface()
+
+	case reflect.Chan:
+		// verify if we already have copied that string
```
Comment (ScottMansfield) [inline deepcopy.go]: somewhat confusing comments, i assume it was just a copy/paste error.

File: deepcopy.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+		ch := reflect.MakeChan(typ, value.Len())
+
+		// store the copy in the map for future references
+		c.pointers[unsafe.Pointer(value.Pointer())] = ch.Interface()
+
+		return ch.Interface()
+
+	case reflect.Map:
+		// verify if we already have copied that string
```
Comment (ScottMansfield) [inline deepcopy.go]: comment

File: deepcopy.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	case reflect.UnsafePointer:
+		return unsafe.Pointer(value.Pointer())
+
+	case reflect.String:
+		// ensure a proper copy of the strings content
+		src := []byte(value.String())
+		return string(src)
+
+	case reflect.Func:
+		// func cannot be copied per say, that means closures will remain closure.
```
Comment (ScottMansfield) [inline deepcopy.go]: this is extremely nitpick-y but it's "per se" not "per say"

File: deepcopy.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+		return m.Interface()
+
+	case reflect.Struct:
+		s := reflect.Indirect(reflect.New(typ))
+		for i := 0; i < value.NumField(); i++ {
+			Sudo(s.Field(i)).Set(reflect.ValueOf(c.deepCopy(Sudo(value.Field(i)).Interface())))
+		}
+		return s.Interface()
+	case reflect.Ptr:
+		// verify if we already have copied that string
```
Comment (ScottMansfield) [inline deepcopy.go]: comment

File: deepcopy.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+		return float32(value.Float())
+	case reflect.Float64:
+		return float64(value.Float())
+
+	case reflect.Complex64: // complexs
+		return complex64(value.Complex())
+	case reflect.Complex128:
+		return complex128(value.Complex())
+
+	case reflect.Uintptr:
```
Comment (ScottMansfield) [inline deepcopy.go]: In the documentation for the function you mention it doesn't copy these, I'd like to see some comments saying so here as well.

File: deepcopy.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+		interf := ptr.Interface()
+		// store the copy in the map for future references
+		c.pointers[unsafe.Pointer(value.Pointer())] = interf
+
+		cp := reflect.ValueOf(c.deepCopy(reflect.Indirect(value).Interface()))
+		reflect.Indirect(ptr).Set(cp)
+
+		return interf
+	case reflect.Interface:
+		panic("Is that even reacheable ?")
```
Comment (ScottMansfield) [inline deepcopy.go]: This ought to be more descriptive so if it does get reached it's not a mystery what triggered it.

File: deepcopy_test.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
+package dark
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestDeepCopy(t *testing.T) {
+	tests := []struct {
```
Comment (ScottMansfield) [inline deepcopy_test.go]: I don't see a test verifying the unsafe.Pointer is *not* copied

File: deepcopy_test.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		s0: s,
+		s1: s[:5],
+	}
+	u := DeepCopy(r)
+	if !reflect.DeepEqual(u, r) {
+		t.Fatalf("not equal got %v, want %v", u, r)
+	}
+}
+
+// TestDeepCopyChannel tests the copying of channels, reflect.DeepEqual doesn't
```
Comment (ScottMansfield) [inline deepcopy_test.go]: You could add a test with a buffered channel, too, to check if the contents are copied over.

File: deepcopy_test.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+	i := deepInterface((*deepCopier)(nil))
+	cp := DeepCopy(i)
+	if reflect.TypeOf(i) != reflect.TypeOf(cp) {
+		t.Fatalf("types are expected to be the same, %T != %T", i, cp)
+	}
+}
+
+// TestDeepCopyRing tests that a circular reference doesn't go out of control
+// and allocate everything.
+func TestDeepCopyRing(t *testing.T) {
```
Comment (ScottMansfield) [inline deepcopy_test.go]: this is a great idea for a test.

**Comment by hydroflame** [comment]
@ScottMansfield I've addressed most of your comments and good catch, there was a bug with unsafe-like pointers. However I have not fixed the copying of channel content as it's kindof a mystery as to how this would even be done. channel are intrinsics to the compiler and it would be pretty unstable to try to copy the content. I have however explicitly spelled that out in the comment for the function

**Pull request merged**